src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/ICastableHelpers.cs

```diff
@@ -22,5 +23,30 @@ internal static RuntimeType GetImplType(ICastable castable, RuntimeType interfac
         {
             return castable.GetImplType(new RuntimeTypeHandle(interfaceType)).GetRuntimeType();
         }
+
+        [Diagnostics.StackTraceHidden]
+        internal static RuntimeType? GetInterfaceImplementation(ICastableObject castableObject, RuntimeType interfaceType, bool throwIfNotFound)
+        {
+            RuntimeTypeHandle handle = castableObject.GetInterfaceImplementation(new RuntimeTypeHandle(interfaceType), throwIfNotFound);
+            if (handle.Equals(default))
+            {
+                if (throwIfNotFound)
+                    throw new InvalidCastException(SR.Format(SR.InvalidCast_FromTo, castableObject.GetType().ToString(), interfaceType.ToString()));
+
+                return null;
+            }
+
+            RuntimeType implType = handle.GetRuntimeType();
+            if (!implType.IsInterface)
+                throw new InvalidProgramException(SR.Format(SR.ICastableObject_NotInterface, implType.ToString()));
+
+            if (implType.GetCustomAttribute<CastableObjectImplementationAttribute>() == null)
```

Comment:
`IsDefined` is cheaper. It does not need to create instance of the attribute.

Reply:
(I understand that it is looks nice to do these checks via reflection in C#, but it is significantly more expensive than doing the same using unmanaged APIs on the VM side. Something to watch for.)